### PR TITLE
Add a PR-time check for non-standard NPM registry

### LINKED ISSUE #2876: Check for non-standard NPM registry in package-lock.json as a step in CI
Having references to the Red Hat internal npm repository in the package-lock.json prevents non-Red Hat developers from installing the vscode-java npm dependencies. As mentioned in https://github.com/redhat-developer/vscode-java/issues/2874#issuecomment-1380478948, npm will always use the configured repository to download packages, regardless of what's in the package-lock.json, although it will attempt to validate cached packages against what's in the package-lock.json.

It would be nice to check for repository.engineering.redhat.com in package-lock.json as a part of the CI so that in the future we don't accidentally commit changes like this.

### PR DESCRIPTION
**Pull request by fsouza**
This should prevent issues when people outside of Red Hat try to contribute to the project.

Using JavaScript instead of shell script just in case we want to support development in environments where shell script is tricky.

Closes #2876.

### CODE REVIEW

**Comment by fsouza** [comment]
Example failure: https://github.com/fsouza/vscode-java/actions/runs/5160457538/jobs/9296530062

One thing that bothers me is that the error isn't super explicit, but it's there. Should we add some all cap words? heh

**Comment by rgrunber** [comment]
Would it be possible to move the check/fix logic into [`gulpfile.js`](https://github.com/redhat-developer/vscode-java/blob/master/gulpfile.js) ? 

Maybe something like :

```js
gulp.task('repo_check', function (done) {
	repo_check_fn();
	done();
});

function repo_check_fn(){
... the repo check logic ...
}
```

In the `scripts` section of the package.json, you can then call it with :

```json
"repo:check": "./node_modules/.bin/gulp repo_check"
```

**Comment by fsouza** [comment]
Done, moved to gulp. Here's an example of it working:

```
％ npm run repo:check

> java@1.20.0 repo:check
> ./node_modules/.bin/gulp repo_check

[23:08:36] Using gulpfile ~/Projects/os/3/vscode-java/gulpfile.js
[23:08:36] Starting 'repo_check'...
[23:08:36] 'repo_check' errored after 1.55 ms
[23:08:36] Error: Found references to repository.engineering.redhat.com in the file package-lock.json. Please fix it with 'npm run repo:fix'
    at /Users/fsouza/Projects/os/3/vscode-java/gulpfile.js:227:8
    at repo_check (/Users/fsouza/Projects/os/3/vscode-java/node_modules/undertaker/lib/set-task.js:13:15)
    at bound (domain.js:421:15)
    at runBound (domain.js:432:12)
    at asyncRunner (/Users/fsouza/Projects/os/3/vscode-java/node_modules/async-done/index.js:55:18)
    at processTicksAndRejections (internal/process/task_queues.js:77:11)
％ npm run repo:fix

> java@1.20.0 repo:fix
> ./node_modules/.bin/gulp repo_fix

[23:08:39] Using gulpfile ~/Projects/os/3/vscode-java/gulpfile.js
[23:08:39] Starting 'repo_fix'...
successfully fixed package-lock.json
[23:08:39] Finished 'repo_fix' after 3.27 ms
％ npm run repo:check

> java@1.20.0 repo:check
> ./node_modules/.bin/gulp repo_check

[23:08:43] Using gulpfile ~/Projects/os/3/vscode-java/gulpfile.js
[23:08:43] Starting 'repo_check'...
[23:08:43] Finished 'repo_check' after 1.74 ms

```

File: gulpfile.js [hunk 1]
```diff
@@ -14,6 +14,10 @@ const originalTestFolder = path.join(__dirname, 'test', 'resources', 'projects',
 const tempTestFolder = path.join(__dirname, 'test-temp');
 const testSettings = path.join(tempTestFolder, '.vscode', 'settings.json');
 const JDT_LS_SNAPSHOT_URL = "http://download.eclipse.org/jdtls/snapshots/jdt-language-server-latest.tar.gz"
+const INTERNAL_RH_REPOSITORY_RE = new RegExp(
+	String.raw`"resolved":\s*"https://repository.engineering.redhat.com/nexus/repository/`,
```
Comment (rgrunber) [inline gulpfile.js:18]: Hey, sorry for the delay on this. Just one final thing I was thinking about. As it's generally not a good practice to expose internals (and for the sake of making this more generic) could we change the matching logic to something like :

`resolved":\s*"https://(.+)registry.npmjs.org/`

In other words we could detect if `https://registry.npmjs.org` is never matched, and if so remove the capture group from the above regular expression.

Comment (fsouza) [reply gulpfile.js:18]: Sounds good, I'll rename the variable too!

**Comment by fsouza** [comment]
Generalized the regexp, here's a sample failure: https://github.com/fsouza/vscode-java/actions/runs/5365029176/jobs/9733582822 (the CI job ran before I fixed the error message, see below for the failure with the updated error message)

Also locally:

```
％ npm run repo:check

> java@1.20.0 repo:check
> ./node_modules/.bin/gulp repo_check

[11:34:45] Using gulpfile ~/Projects/os/3/vscode-java/gulpfile.js
[11:34:45] Starting 'repo_check'...
[11:34:45] 'repo_check' errored after 986 μs
[11:34:45] Error: Found references to the internal registry in the file package-lock.json. Please fix it with 'npm run repo:fix'
    at /Users/fsouza/Projects/os/3/vscode-java/gulpfile.js:227:8
    at taskWrapper (/Users/fsouza/Projects/os/3/vscode-java/node_modules/undertaker/lib/set-task.js:13:15)
    at bound (node:domain:433:15)
    at runBound (node:domain:444:12)
    at asyncRunner (/Users/fsouza/Projects/os/3/vscode-java/node_modules/async-done/index.js:55:18)
    at process.processTicksAndRejections (node:internal/process/task_queues:77:11)
％ npm run repo:fix

> java@1.20.0 repo:fix
> ./node_modules/.bin/gulp repo_fix

[11:34:56] Using gulpfile ~/Projects/os/3/vscode-java/gulpfile.js
[11:34:56] Starting 'repo_fix'...
successfully fixed package-lock.json
[11:34:56] Finished 'repo_fix' after 2.18 ms
％ npm run repo:check
％ git diff package-lock.json
diff --git a/package-lock.json b/package-lock.json
index 37d7b92..a06e438 100644
--- a/package-lock.json
+++ b/package-lock.json
@@ -525,7 +525,7 @@
     },
     "@types/sinon": {
       "version": "10.0.13",
-      "resolved": "https://repository.engineering.redhat.com/nexus/repository/registry.npmjs.org/@types/sinon/-/sinon-10.0.13.tgz",
+      "resolved": "https://registry.npmjs.org/@types/sinon/-/sinon-10.0.13.tgz",
       "integrity": "sha512-UVjDqJblVNQYvVNUsj0PuYYw0ELRmgt1Nt5Vk0pT5f16ROGfcKJY8o1HVuMOJOpD727RrGB9EGvoaTQE5tgxZQ==",
       "dev": true,
       "requires": {
@@ -1122,7 +1122,7 @@
     },
     "ansi-styles": {
       "version": "2.2.1",
-      "resolved": "https://repository.engineering.redhat.com/nexus/repository/registry.npmjs.org/ansi-styles/-/ansi-styles-2.2.1.tgz",
+      "resolved": "https://registry.npmjs.org/ansi-styles/-/ansi-styles-2.2.1.tgz",
       "integrity": "sha512-kmCevFghRiWM7HB5zTPULl4r9bVFSWjz62MhqizDGUrq2NWuNMQyuv4tHHoKJHs69M/MF64lEcHdYIocrdWQYA==",
       "dev": true
     },
@@ -1724,7 +1724,7 @@
     },
     "chalk": {
       "version": "1.1.3",
-      "resolved": "https://repository.engineering.redhat.com/nexus/repository/registry.npmjs.org/chalk/-/chalk-1.1.3.tgz",
+      "resolved": "https://registry.npmjs.org/chalk/-/chalk-1.1.3.tgz",
       "integrity": "sha512-U3lRVLMSlsCfjqYPbLyVv11M9CPW4I728d6TCKMAOJueEeB9/8o+eSsMnxPJD+Q+K909sdESg7C+tIkoH6on1A==",
       "dev": true,
       "requires": {
@@ -2456,7 +2456,7 @@
         },
         "string_decoder": {
           "version": "0.10.31",
-          "resolved": "https://repository.engineering.redhat.com/nexus/repository/registry.npmjs.org/string_decoder/-/string_decoder-0.10.31.tgz",
+          "resolved": "https://registry.npmjs.org/string_decoder/-/string_decoder-0.10.31.tgz",
           "integrity": "sha512-ev2QzSzWPYmy9GuqfIVildA4OdcGLeFZQrq5ys6RtiuF+RQQiZWr8TZNyAcuVXyQRYfEO+MsoB/1BuQVhOJuoQ==",
           "dev": true
         }
@@ -2640,7 +2640,7 @@
     },
     "escape-string-regexp": {
       "version": "1.0.5",
-      "resolved": "https://repository.engineering.redhat.com/nexus/repository/registry.npmjs.org/escape-string-regexp/-/escape-string-regexp-1.0.5.tgz",
+      "resolved": "https://registry.npmjs.org/escape-string-regexp/-/escape-string-regexp-1.0.5.tgz",
       "integrity": "sha512-vbRorB5FUQWvla16U8R/qgaFIya2qGzwDrNmCZuYKrbdSUMG6I1ZCGQRefkRVhuOkIGVne7BQ35DSfo1qvJqFg==",
       "dev": true
     },
@@ -4173,7 +4173,7 @@
     },
     "has-ansi": {
       "version": "2.0.0",
-      "resolved": "https://repository.engineering.redhat.com/nexus/repository/registry.npmjs.org/has-ansi/-/has-ansi-2.0.0.tgz",
+      "resolved": "https://registry.npmjs.org/has-ansi/-/has-ansi-2.0.0.tgz",
       "integrity": "sha512-C8vBJ8DwUCx19vhm7urhTuUsr4/IyP6l4VzNQDv+ryHQObW3TTTp9yB68WpYgRe2bbaGuZ/se74IqFeVnMnLZg==",
       "dev": true,
...
```

**Pull request merged**